### ISSUE TITLE
Add MSPatchInfo to the file spec

### ISSUE THREAD

**Issue by opsGavin** [sketch-hq/sketch-file-format#94]
The contents go like this:
```
MSPatchInfo {
  lastIntegratedPatchID: string
  statusInfo: MSCollaborationStatusInfo {
    status: string
    uploadedPatches: array of strings
    isConsolidating: bool
    shareID: string
    baseVersionID: string
    versionID: string
  }
  MSPatchContainer {
    localPatches: an array of MSPatch objects
  }
```

This object contains the information we need to track the state of collaboration documents and should really only be present in documents within the hidden cloud folder on users machines.

The contents of MSPatch objects are described in [Notion](https://www.notion.so/sketchapp/Overview-6f3f6dd858e74dd084c08e6b5e24b4b6#d95b4faaaa434714826d6f10d97efd6e)

@jedrichards hope that's useful. I'm not sure how all this stuff works, so ping me if you need more info.

**Comment by jedrichards** [comment]
Thanks @opsGavin! What would be even better is if you could also dump some JSON here, as it appears on disk? Since that's the data that is being described by the schemas in this repo. I note that `_class` strings like `MSImmutablePatchInfo` appear on disk, but are not mentioned in your description above.

**Comment by opsGavin** [comment]
Sorry @jedrichards I'm not ignoring this. We're looking at making some changes here, but if you bear with me I'll come back and add some examples when that's done.

**Comment by jedrichards** [comment]
No problemo, sounds good 👍

**Comment by jedrichards** [comment]
Wondering if this is no longer relevant now we've added `documentState`?

**Comment by opsGavin** [comment]
We're still going to need this for co-edit in some form, but I'm not certain how it will look in the end.

**Comment by christianklotz** [comment]
@opsGavin is this still relevant?

**Comment by opsGavin** [comment]
Yes - still relevant and probably a little more stable now. Here's an example of the patchInfo from a document I have here.

```
  "patchInfo" : {
    "lastIntegratedPatchID" : "f6722512-1c0c-4a21-b016-31e84762652c",
    "status" : {
      "statusString" : "COLLABORATING",
      "_class" : "SharedEditing.MSCoEditStatusInfo",
      "shareID" : "30aa41e9-f735-4696-b005-263a8df6da47"
    },
    "_class" : "MSImmutablePatchInfo",
    "localPatches" : [
      {
        "_ref_class" : "MSPatch",
        "_class" : "MSJSONFileReference",
        "_ref" : "patches\/01d7c53c-d446-4371-ae84-6b023327d659"
      },
      {
        "_ref_class" : "MSPatch",
        "_class" : "MSJSONFileReference",
        "_ref" : "patches\/9f759e56-349f-40ae-8ae4-b3b4fd1cbdcc"
      },
      {
        "_ref_class" : "MSPatch",
        "_class" : "MSJSONFileReference",
        "_ref" : "patches\/a6e9bb74-00e3-46d8-acfc-d2f295b115ca"
      }
    ]
  },
```

The `status` contains information about the document which is basically just the shareID along with a string describing whether a connection was established. I wouldn't expect to see anything other than `COLLABORATING` here although I think `LOCAL` or `INITIAL` may be possible in theory (/cc @mkeiser).

`localPatches` contains references to patch files which have still to be uploaded to the server. Normally there won't be any, but if the user closes the document while off-line there could be and these will be sent when they re-open the document.

Details of the contents of the patch files can be found at https://www.notion.so/sketchapp/Overview-6f3f6dd858e74dd084c08e6b5e24b4b6#d95b4faaaa434714826d6f10d97efd6e. I believe there is also a schema for them somewhere - @mz2 would probably be able to point you to that.

It's worth just re-iterating that this information should only be found in cloud documents which should be hidden from the user and that we certainly don't want anyone trying to edit any of this information.

Shout if you have any questions

**Comment by jedrichards** [comment]
Still feels like we haven't come to a decision on this one. Since `patchInfo` is by all accounts something of an implementation detail, does it even belong in the public schemas?

**Comment by christianklotz** [comment]
One of the mentioned outcomes was to document it here, not because it's useful for third parties (it isn't) but to remain documented.

**Comment by jedrichards** [comment]
I'd like to unblock this if possible. If we want to go ahead with it there are still one or two todos on the PR, e.g. https://github.com/sketch-hq/sketch-file-format/pull/115#pullrequestreview-488556820.

That said, I'd still lean towards perhaps closing it instead. I understand the desire to document it, but since it isn't really a public-facing property, I assume it's documented already where it needs to be - in the Sketch code base? Does it make sense to document it here too, on the public schema repo, where it could easily get out of date etc?

**Comment by jedrichards** [comment]
@christianklotz This has arrived in base now, so someone could take a look at it, but still seems to the way forwards isn't that clear.

I'm still leaning towards closing it to be honest, but are you are keen to press ahead and document the collab patch format in the public schemas?

Although closed, the schema changes in the linked [PR](https://github.com/sketch-hq/sketch-file-format/pull/115/files) are probably still relevant, although would be worthwhile to get some Mac dev 👀 on them to see if they are still correct.

**Comment by mkeiser** [comment]
I think this is the version that we shipped in 71 (cc @opsGavin to double check):

```
    "patchInfo": {
        "_class": "MSImmutablePatchInfo",
        "baseVersionID": "72704274-de94-4c29-b04a-984694d652f0",
        "lastIntegratedPatchID": "1a8f3c2a-8b32-4f78-8d75-3ecbf2d2c07a",
        "localPatches": [
            {
                "_class": "MSJSONFileReference",
                "_ref": "patches/151bdc13-f065-4772-8299-32e92caf05c5",
                "_ref_class": "MSPatch"
            }
        ],
        "receivedPatches": [
            {
                "_class": "MSJSONFileReference",
                "_ref": "patches/6d1dd1b4-9323-42d2-9670-e2b57a8e52df",
                "_ref_class": "MSPatch"
            }
        ]
    }
```
